src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs

```diff
@@ -587,9 +587,17 @@ private bool BatchesNeedProcessing
 
                     if (_nonGreedyState != null)
                     {
-                        // We can make a triggered batch using postponed messages
-                        if (_nonGreedyState.AcceptFewerThanBatchSize &&
-                            (_messages.Count > 0 || (_nonGreedyState.PostponedMessages.Count > 0 && boundedCapacityAvailable > 0)))
+                        // If a batch was triggered and we have any messages, we can create a batch from what we already have.
+                        if (_nonGreedyState.AcceptFewerThanBatchSize && _messages.Count > 0)
+                            return true;
+
+                        // At this point, to make a batch we'll need to consume postponed messages, but we can't do
+                        // that if we're declining all future messages.
+                        if (_decliningPermanently)
```

Comment:
>if [](start = 24, length = 2)

is it possible between this check and the previous check, some message get added?

Reply:
It shouldn't be.  Above we assert that we're holding the "incoming lock":
https://github.com/dotnet/runtime/blob/4cd92ba704e3ca4bc3b5470ec38e31ecc12995d3/src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L571
which is the lock that's taken when messages are offered to the block:
https://github.com/dotnet/runtime/blob/4cd92ba704e3ca4bc3b5470ec38e31ecc12995d3/src/libraries/System.Threading.Tasks.Dataflow/src/Blocks/BatchBlock.cs#L366